Add script formatter, comment parsing in parser, and editor settings

lazzy-cipher opened the pull request:
This PR builds upon and finalizes the work done in [godotengine/godot/pull/55835](https://github.com/godotengine/godot/pull/55835).

This PR introduces a GDScript formatter in Godot's script editor. Users can access it through Edit/Format Code, <kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd>, or by enabling the new *Format On Save* editor setting while saving their work.

Integrating a formatter in Godot's script editor improves it as an IDE for GDScript. Additionally, the formatter will improve developer's adherence to the [official GDScript style guide](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html).

We encourage users to test the formatter on large code bases in order to detect any quirks or bugs that need to be addressed.

Additionally, this PR includes support for comments through a header and inline. During parsing, each comment of a comment block are added to a comment header vector. The next node (variable, function, annotation) has this vector set as the header comments. The following comment until a newline, if present, is set as the inline comment.

The commits have been kept separate for ease of review, but they will be squashed before merging. Lastly, there are no new dependencies included in this PR.

_Production edit: Closes https://github.com/godotengine/godot-proposals/issues/3630_

Razoric480 committed:
Add script formatter, comment parsing in parser, and editor settings

* Add verbose messages for debugging purposes
* Add disabled lines handling to parser
* Force newline at end of code

Signed-off-by: Adam Scott <ascott.ca@gmail.com>

YuriSizov commented:
Thanks for looking into this!

> The commits have been kept separate for ease of review

This PR contains only one commit.

NathanLovato commented:
Just tested it on a number of files, here are snippets causing unexpected results:

```gdscript
func _ready() -> void:
	energy_slider.value_changed.connect(func(value):
		light.energy = energy_slider.value)
```

Gets formatted into this:

```gd
func _ready() -> void:
	energy_slider.value_changed.connect(func(value):
		light.energy = energy_slider.value
)
```

This happens differently as you duplicate the two lines with the function literal.

This:

```gdscript
func _ready() -> void:
	energy_slider.value_changed.connect(func(value):
		light.energy = energy_slider.value)

	height_slider.value_changed.connect(func(value):
		light.height = height_slider.value)

	height_slider.value_changed.connect(func(value):
		light.height = height_slider.value)
```

Alternates between the following and the starting state when running the formatter.

```gdscript
func _ready() -> void:
	energy_slider.value_changed.connect(func(value):
		light.energy = energy_slider.value
)

	height_slider.value_changed.connect(func(value):
		light.height = height_slider.value
)

	height_slider.value_changed.connect(func(value):
		light.height = height_slider.value)
```

NathanLovato commented:
Longer annotations get placed on their own lines, unlike shorter ones. Not a bug per say, but I'm not 100% sure if it's intentional.

```gdscript
@export_multiline var text := ""
@export var logo_visible := false
```

Becomes:

```gd
@export_multiline
var text := ""
@export var logo_visible := false
```

NathanLovato commented:
Comment inside a function gets indented based on previous block.

```gdscript
func popup(immediate := false) -> void:
	if not is_inside_tree():
		await ready
	
	# Force container to be its smallest size possibles
	size = Vector2.ZERO
```

Becomes:

```gdscript
func popup(immediate := false) -> void:
	if not is_inside_tree():
		await ready

		# Force container to be its smallest size possibles
	size = Vector2.ZERO
```

NathanLovato commented:
Last line in a setter function attached to a property definition gets pushed down up to two times if following an ending code block.

```gdscript
var text := "":
	set(value):
		text = value
		if not is_inside_tree():
			await ready
		rich_text_label.text = text
```

First becomes

```gdscript
var text := "":
	set(value):
		text = value
		if not is_inside_tree():
			await ready

		rich_text_label.text = text
```

And after a second run of the formatter, it becomes

```gdscript
var text := "":
	set(value):
		text = value
		if not is_inside_tree():
			await ready


		rich_text_label.text = text
```

After that it's stable with subsequent formatter runs.

NathanLovato commented:
Arrays struggle to wrap when combined with method calls currently. Here's a very long line to generate a random nickname:

```gdscript
var nickname: String = ["Brave", "Horrific", "Courageous", "Terrific", "Fair", "Conqueror", "Victorious", "Glorious", "Invicible"].pick_random() + ["Leopard", "Cheetah", "Bear", "Turtle", "Rabbit", "Porcupine", "Hare", "Pigeon", "Albatross", "Crow" ].pick_random()
```

The formatter doesn't manage to wrap it to stay below the max line length. It produces this:

```gdscript
var nickname: String = (
		["Brave", "Horrific", "Courageous", "Terrific", "Fair", "Conqueror", "Victorious", "Glorious", "Invicible"].pick_random(
			)
		+ ["Leopard", "Cheetah", "Bear", "Turtle", "Rabbit", "Porcupine", "Hare", "Pigeon", "Albatross", "Crow"].pick_random(
			)
)
```

Not a super important case, but each array is longer than max line length, so I'd expect them to be wrapped vertically, as the formatter does well when all you have is an array (without the method call). E.g. this

```gdscript
var adjectives := ["Brave", "Horrific", "Courageous", "Terrific", "Fair", "Conqueror", "Victorious", "Glorious", "Invicible"]
```

Becomes the following, as expected:

```gdscript
var adjectives := [
	"Brave",
	"Horrific",
	"Courageous",
	"Terrific",
	"Fair",
	"Conqueror",
	"Victorious",
	"Glorious",
	"Invicible",
]
```

anvilfolk commented:
I've been working with GDScript docstrings lately, and from a super quick look through the code, that feels relatively similar to this idea that comments get assigned to relevant members. Is the purpose of that assignment to be able to format comments associated with members?

Because if it is, I wonder whether there's some refactoring possible there. Docstrings can be considered a special case of comment (they use `##` instead of a single `#`). Either way, that refactoring should be done in a separate PR. I am just trying to figure out whether this is being used for documentation at all, and might thus be duplicated some work already done for docstrings!

monkeez commented:
Will it be possible to call the formatter via the CMD line? Main use case would be if working with an external editor.

NathanLovato commented:
> Will it be possible to call the formatter via the CMD line? Main use case would be if working with an external editor.

For external editors, the first goal should be to add formatting to the language server, which will instantly provide support for most editors (vscode, neovim, jetbrain, emacs, ...). That's definitely something to add once the formatter gets merged. 

Command line support can and should probably be added for use in continuous integration.

ajreckof commented:
I've tested it a bit on a few of my codes here are what i found.

I'm not sure the formatter should change annotation placement, I feel it should keep how you had it, especially since there is no explicit guideline on whether to put it in its separate line or not. 

large number are wrongly formatted, if they are not formatted they do not change and if they are formatted every `_` gets removed. See here for proper formatting of large numbers
<img width="856" alt="Capture d’écran 2023-04-24 à 07 10 45" src="https://user-images.githubusercontent.com/66184050/233905138-5e8c53eb-ac58-41b2-9d66-6398657a0416.png">

when formatting
```
printt("a lot of arguments", "a lot of arguments", "a lot of arguments", "a lot of arguments", "a lot of arguments")
```
it produces this 
```
printt(
			"a lot of arguments",
			"a lot of arguments",
			"a lot of arguments",
			"a lot of arguments",
			"a lot of arguments"
	)
```
which is missing the trailing coma (only tested with function but the trailing comma might be misiing in arrays and dictionaries too )

Moreover when separating into multiple line as in the last example if the line is just after a unindent than it will put two empty lines instead of the one needed. reformating will then reduce to one line.

I also had a similar issue to the one reported here https://github.com/godotengine/godot/pull/76211#issuecomment-1513655346 but it was myself with nested functions 

when having multiple unindent you will have as much empty line as the number of unindent which might result in the next line being a 4 line after (strangely this does not happen in the last function of the script)

RolandMarchand commented:
@anvilfolk Thank you very much for the feedback!

For the annotations, `@tool` and `@icon` are applied to the parser itself, similar to a flag. There is no information in the parser as to which one came first. Other annotations do have a stack system, though. We could implement a system to keep track of those two annotations' order, but I do not think it's worth the effort.

Aside from that, yes, there are still some bugs. Thank you so much for the screenshots, I will make test cases out of them and fix them ASAP :+1:

dalexeev commented:
> large number are wrongly formatted, if they are not formatted they do not change and if they are formatted every `_` gets removed.

The exact formatting of the literals is lost:

```diff
-    print(0xFFFF)
-    print(1.2e-5)
-    print(0.1111111111111133) # <- Out of precision.
-    print("\t \n \uFFFD")
-    print(""" key="value" """)
+    print(65535)
+    print(0.000012)
+    print(0.11111111111111) # <- Out of precision.
+    print("	 
+ �")
+    print(' key="value" ')
```

Possible solution: when allocating a `LiteralNode`, copy the `Token::source` property into a new `LiteralNode` property and use it in the formatter.

I'm a bit worried about this approach (restoring source code from AST) but overall the code looks good to me, excellent work!

Now we need to make sure that the parser has all the necessary information, including the order of the elements. For example, now the formatter moves the setter before the getter, but we do not have this recommendation in the style guide (and in the GDScript reference examples, the getter is placed before the setter). I think the formatter should not change anything that is not regulated by the style guide.

aaronfranke commented:
@dalexeev We should consider changing the recommended order of the class doc comment to be like this:

```gdscript
## Class doc comment.
class_name MyClass
extends Node
```

Every other doc comment, and all annotations, appear above/before what they describe, not after.

To avoid breaking compat, we could do this in steps, first support both orders for one Godot release (like 4.1), then enforce it being above for the next release (like 4.2).

dalexeev commented:
@aaronfranke You are probably right, although both `class_name` and `extends` are optional (if `extends` is omitted, the class extends `RefCounted`).

lazzy-cipher committed:
Duplicate tool annotation error handling moved to parser in tests

Signed-off-by: Adam Scott <ascott.ca@gmail.com>

lazzy-cipher committed:
Fixed isolated parenthesis after lambda at the end of suite, implemented test case for it

Signed-off-by: Adam Scott <ascott.ca@gmail.com>

adamscott commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -1292,18 +1297,21 @@ void ScriptEditor::_menu_option(int p_option) {
+						}
 					}
+				} else {
+					current->format_code();
```
I know that you only wrapped existing code with an `if` condition, but as this code is repeating itself multiple time in the file, maybe you could create a function that "formats" the code.

adamscott commented on modules/gdscript/gdscript_format.cpp:
```diff
@@ -0,0 +1,2194 @@
+		}
+		output += "@icon(\"";
+		output += p_node->icon_path;
+		output += "\")";
```
You could simplify the code by using `vformat()`.
```cpp
output += vformat("@icon(\"%s\")", p_node->icon_path)
```

adamscott replied:
No, nevermind. Finally, vformat is way less performant than `vformat()`.

adamscott committed:
Add `ScriptEditor::_format_code()` to remove repeating code throughout the file

Signed-off-by: Adam Scott <ascott.ca@gmail.com>

adamscott committed:
Fix literals being parsed when formatting code

adamscott commented:
As this is a HUGE PR, I suggest to merge it sooner than later. As there's multiple issues that have been underlined in the comments, I think it will be easier to solve them in chunks with each of them having their own issue tracked rather than trying to address them all in this PR.

So, I could rebase the commits to create one and we could merge the PR. 

@akien-mga @YuriSizov @vnen I think it would be a great addition to 4.1, especially as this will let us time to fix the issues before and after the feature freeze.

adamscott commented:
> I'm a bit worried about this approach (restoring source code from AST) but overall the code looks good to me, excellent work!

@dalexeev The popular code formatter Prettier for NodeJS uses the AST generated by Babel to "prettify" Javascript and Typescript code. So this is a valid way to format code.

adamscott reviewed (dismissed):
Terrific work, @RolandMarchand. Thank you very much! As I said previously, even if it's not perfect, I think your code formatter is stable enough to be merged for 4.1.

dalexeev commented:
> As this is a HUGE PR, I suggest to merge it sooner than later.

If there are no regressions found in existing features (what about doc comments?), then I support this. We can warn users that formatting isn't working perfectly, but since it's off by default I don't think it's a problem, users can just not use it and nothing will change for them.

adamscott commented:
Currently working on making the build pass.

adamscott committed:
Use variable instead of accessor

adamscott committed:
Trying to make the build script pass

adamscott committed:
Revert adding a trailing comma for function calls

... as it breaks the tests.

NathanLovato commented:
I second a merge soon: the formatter works fine overall - remaining issues will be edge cases in code that it formats without errors or crashes, but that don't have ideal formatting in the output.

There are certainly many more to find, and finding these will involve a lot of testing by the community, running the formatter on their code. And this won't happen unless the formatter's available in a build.

dalexeev commented:
What about points 3 and 4 in [the comment](https://github.com/godotengine/godot/pull/76211#issuecomment-1519490403)? I can test this PR again when you confirm you are done.

adamscott commented:
@dalexeev Working on point 3, the regression of the missing `pass`

aaronfranke commented:
Once everyone else's points are addressed, I can give this a test on The Mirror's codebase before merging. It's likely that I will find a bunch of issues and edge cases testing on our codebase with about 50k lines of GDScript ;)

adamscott committed:
Fix functions without statements

adamscott commented:
> Once everyone else's points are addressed, I can give this a test on The Mirror's codebase before merging. It's likely that I will find a bunch of issues and edge cases testing on our codebase with about 50k lines of GDScript ;)

@aaronfranke Feel free to test The Mirror's codebase :fire:, but as I said:

> As this is a HUGE PR, I suggest to merge it sooner than later. As there's multiple issues that have been underlined in the comments, I think it will be easier to solve them in chunks with each of them having their own issue tracked rather than trying to address them all in this PR.

adamscott commented:
> @dalexeev Working on point 3, the regression of the missing `pass`

Done.

aaronfranke commented:
@adamscott @NathanLovato I disagree with the "merge now, fix later" mentality. This is not just some optional feature that runs on top of GDScript ("off by default" as @dalexeev puts it), it also includes changes to the GDScript parser, which can break scripts if it's not tested.

I tested it and I found a few parser errors (without formatting), and many issues with the formatter. Some of the formatter issues I've noted below are the same issues as the ones I pointed out before, so I guess the test cases I posted about before were not re-tested with the latest version of the PR.

Side note, I would greatly appreciate the ability to selectively toggle the formatter off for specific sections of the code. There are 2 use cases here. One is that the formatter may make a mistake, so you want to turn it off for that part. The other is that sometimes the most readable way to write something may be something we don't recommend, such as having many lines that exceed the length limit but are aligned with each other so it's very easy to read.

---

This code errors with the message "Expected end of statement after variable declaration, found "Dedent" instead."

```gdscript
func _ready():
	var v := Vector2()
	var a = v[0] # A
```

---

This code errors with the message "Expected expression as dictionary key."

```gdscript
func _ready():
	var dict = {
		"key1": "value1", # Comment
		# Comment
		"key2": "value2",
	}
```

---

The game does not run, Godot says "Parser Error: Could not resolve class ..." for some scripts and highlights the extends line in red. I suspect it's related to this error spammed in the console:

```
ERROR: Parser bug: Mismatch in extents tracking stack.
   at: complete_extents (modules/gdscript/gdscript_parser.cpp:4751)
ERROR: Parser bug: Mismatch in extents tracking stack.
   at: complete_extents (modules/gdscript/gdscript_parser.cpp:4751)
```

---

Unique nodes lose their special syntax and are converted to using the `$` operator. This:

```gdscript
@onready var unique_node = %UniqueNode
```

becomes this:

```gdscript
@onready var unique_node = $"%UniqueNode"
```

---

This code does not format right, this:

```gdscript
var _MAX_FILE_SIZE = 1 << 20 # 1 MiB
```

becomes this:

```gdscript
var _MAX_FILE_SIZE = (
		1
		<< 20 # 1 MiB
)
```

However, this one is minor, I can just move the comment to the line above to prevent this issue.

---

This code does not format right, this:

```gdscript
func _ready():
	var some_node
	# This bug only happens when the comment on the previous line exceeds 80 chars.
	some_node.visible = 5 > 3
```

becomes this:

```gdscript
func _ready():
	var some_node
	# This bug only happens when the comment on the previous line exceeds 80 chars.
	some_node.visible = (
			5 > 3
	)
```

and probably related, this:

```gdscript
func _ready():
	var some_node
	# This bug only happens when the comment on the previous line exceeds 80 chars.
	some_node.visible = some_node.some_func()
```

becomes this:

```gdscript
func _ready():
	var some_node
	# This bug only happens when the comment on the previous line exceeds 80 chars.
	some_node.visible = some_node.some_func(
			)
```

---

Comments after an indented block are forcefully moved over:

```gdscript
func _ready():
	if true:
		pass
	# Description of A
	var a
```

becomes this:

```gdscript
func _ready():
	if true:
		pass

		# Description of A
	var a
```

---

The spacing around the equal sign in function parameters is broken. This:

```gdscript
func my_func(optional_param = null):
	pass
```

becomes this:

```gdscript
func my_func(optional_param= null):
	pass
```

---

This formatter explodes (and spits out invalid GDScript code) when trying to format this GUID code:

<details>

```gdscript
static func generate_guid() -> String:
	var b = []
	return "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x" % [
			# low
			b[0], b[1], b[2], b[3],
			# mid
			b[4], b[5],
			# hi
			b[6], b[7],
			# clock
			b[8], b[9], b[10], b[11], b[12], b[13], b[14], b[15]
	]
```

becomes this: (btw, pretend var b is actually a 16-element array)

```gdscript
static func generate_guid() -> String:
	var b = []
	return (
			"%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x"
			% # low
	# low
		[
			b[
				0
			],
			b[
				1
			],
			b[
				2
			],
			b[
				3
			],
			# mid
			b[
				4
			],
			b[
				5
			],
			# hi
			b[
				6
			],
			b[
				7
			],
			# clock
			b[
				8
			],
			b[
				9
			],
			b[
				10
			],
			b[
				11
			],
			b[
				12
			],
			b[
				13
			],
			b[
				14
			],
			b[
				15
			],
		]
	)
```

</details>

---

The formatter puts these on one line despite the fact that it exceeds the maximum line length. This:

```gdscript
func _input(input_event: InputEvent) -> void:
	if input_event.is_action(&"some_long_action_name") or \
			input_event.is_action(&"another_action_name"):
		pass
```

becomes this:

```gdscript
func _input(input_event: InputEvent) -> void:
	if input_event.is_action(&"some_long_action_name") or input_event.is_action(&"another_action_name"):
		pass
```

---

The formatter breaks functions inside of inner classes followed by a variable.

```gdscript
class MyInnerClass:
	var hi

	func my_inner_class_function() -> void:
		pass

	func another_one() -> void:
		pass
```

becomes this:

```gdscript
class MyInnerClass:
	var hi

	


func my_inner_class_function() -> void:
		pass


	func another_one() -> void:
		pass
```

---

Comments above static functions are deleted on format:

```gdscript
## This comment will be deleted on format.
static func test():
	pass
```

---

The formatter still breaks this case (multiplying the number then doing string interpolation):

```gdscript
func _ready():
	var number = 1234.567
	var string = "%1.1f k" % (number * 0.001)
```

Becomes this (breaks because it does string interpolation, then it tries to do string * float):

```gdscript
func _ready():
	var number = 1234.567
	var string = "%1.1f k" % number * 0.001
```

---

A comment at the end of a return value is deleted. This:

```gdscript
func test_func() -> bool:
	if true:
		return true # In this case, stop.
	return false
```

becomes this:

```gdscript
func test_func() -> bool:
	if true:
		return true
	return false
```

---

The formatter wants to split up this line, but I don't think it's correct. This:

```gdscript
@export_exp_easing var min_value: float = 0.0
```

becomes this:

```gdscript
@export_exp_easing
var min_value: float = 0.0
```

---

The formatter still goes crazy when you have a comment at the end of an onready var with a get_node call. This:

```gdscript
@onready var a_node = $ANode
@onready var b_node = a_node.get_node(^"BNode") # Test
```

becomes this after several saves:

```gdscript
@onready var a_node = $ANode
@onready
var b_node = a_node.get_node(
		^"BNode"
) # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test # Test
```

---

I'm still getting these weird extra line diffs, but it eludes me when I try to make a minimal test case:

<img width="266" alt="Screenshot 2023-05-13 at 12 19 22 AM" src="https://github.com/godotengine/godot/assets/1646875/f17f997f-113f-42b2-96ac-c23f098e9e41">

Really I would prefer it to just not leave empty lines after indented blocks, I don't like them. But at the very least, the formatter should not leave more than one.

NathanLovato commented:
@aaronfranke thanks for the tests and feedback, and the problem is exactly that: this PR needs extensive testing on many codebases, but new features don´t get the visibility and testing they need until they land in development or beta releases.

Thankfully, you did show up and took the time to test on a big codebase, so thank you kindly for that. That's really helpful.

adamscott committed:
Change `GDScriptFormatter::format()` signature to make it return an error value

adamscott commented:
Just realized that the test suite had no way to detect if the formatter failed to format due to parsing issues. As the formatted always returned the same code submitted if there was an error, if the test used a properly formatted string, it would pass.

Found a new issue like that. The following code has a parser error: `Parse Error: Expected ":" after "match" patterns. (4:10)`
```gdscript
func _ready() -> void:
	var x = 0
	match x:
		0
```

adamscott committed:
Cleanup GDScriptFormatter test file and results

adamscott committed:
Fix issues formatting special node syntax

adamscott commented:
> Unique nodes lose their special syntax and are converted to using the `$` operator. This:
> 
> ```gdscript
> @onready var unique_node = %UniqueNode
> ```
> 
> becomes this:
> 
> ```gdscript
> @onready var unique_node = $"%UniqueNode"
> ```

Fixed

adamscott committed:
Make default values have spaces

Revert test check order

adamscott committed:
Add more tests based on @aaronfranke's comments

adamscott committed:
Fix splitting of annotations with underscores

adamscott committed:
Add test

adamscott committed:
Add GDSCRIPT macro for test file

adamscott commented on modules/gdscript/tests/test_gdscript_formatter.h:
```diff
@@ -0,0 +1,4146 @@
+		CHECK_FORMAT(code, pre_formatted);
+	}
+
+	TEST_CASE("Should maintain type information for typed arrays") {
+		const String code =
+				GDSCRIPT(R"(
+extends Node
+
+@onready var children: Array[Node] = get_children()
+)");
+		const String pre_formatted =
+				GDSCRIPT(R"(
+extends Node
+<*noop*>
+<*noop*>
+@onready var children: Array[Node] = get_children()
+)");
+
+		CHECK_FORMAT(code, pre_formatted);
+	}
```
I added the `GDSCRIPT` macro that improve script readability in the test file. Scripts must now begin with a `\n` when using `GDSCRIPT`.

Take note that the macro replaces the `<*noop*>` tokens with empty strings, that makes the 2+ newlines compatible with "misc/scripts/header_guards.sh"

adamscott committed:
Fix numerous issues

- fix functions in inner classes
- fix issue with last statement being "multiline"

adamscott committed:
Fix "Should not crash with malformed match" test

adamscott committed:
Fix variable assignations with function call returns

adamscott committed:
Fix binary operators rightmost comment

arkology commented:
Is it correct that with this PR only single fixed code style is allowed, there are only 3 style settings, they are editor settings instead of project settings and everyone OK with that?

adamscott committed:
Fix issues with removed parentheses

Introduce new issues about parentheses, unfortunately

adamscott committed:
Fix issue where previous comment line would count in the split of line

adamscott committed:
Add state management in GDScriptTokenizer (push, pop)

adamscott committed:
Fix parser dictionary issues with comments

adamscott committed:
Fix array formatting

- Fix issue where first comment in array would be considered an header
  comment.
- Fix issue where arrays would wrap if their grand-grand-children have a
  comment.

adamscott commented:
@YuriSizov @akien-mga I would rather push to merge it for 4.1 instead of 4.2.

I fixed a lot of @aaronfranke's issues in the last 2 weeks and I think I could continue to polish the formatter until release freeze on June 27th. The remaining test case issues could be resolved this week and the merge could be done really soon. Majority of them are ones related to the change to keep user's custom parentheses groups.

I think this new feature would be a welcomed one for the 4.1 release.

cc. @NathanLovato

NathanLovato commented:
For me, same reasoning as before: this feature needs the wider user base to use it to run it over the less conventional GDScript code and catch edge cases at this point.

This won't happen until it lands in a public beta and people try migrating their projects to the latest version of Godot. For this reason, I expect that the problem will remain the same in the 4.2 dev cycle and that it won't get required testing until any Godot release's beta / bug fixing phase.

dalexeev commented:
<details>
<summary>Background</summary>

![](https://github.com/godotengine/godot/assets/47700418/cd80e768-c446-463d-8ad5-1b0d6ebcf725)

</details>

The following code is formatted

```gdscript
    if true:
        pass
    else:
        if true:
            pass
        else:
            pass
```

as

```gdscript
    if true:
        pass
    elif true:
        pass
    else:
        pass
```

akien-mga commented:
The PR is not yet in a mergeable state (doesn't pass CI, too big commit log, unclear whether the parser breaking bugs have been fixed). As we've reached the feature freeze, it's therefore automatically too late for such a feature to be merged.

It does need wider testing, but this should happen over a full release cycle (4.2 dev builds, if merged shortly after 4.1 is released), instead of doing a last minute feature merge that might introduce regressions and instability in the (by design very short, 1 month) beta phase.

It's not the same as 4.0, we're changing the release cycle to be a 4 month cycle specifically so that we stop merging big feature PRs with potential for game breaking bugs at a time that would jeopardize the release plans. To release 4.1 in early July as we planned, we need to be conservative with feature merged late in the cycle.

NathanLovato commented:
Makes sense, okay so I had misunderstood the release cycles you wanted to have. It's closer to Blender's, where you merge big features early in the cycle, right? All good, let's try to get this merged early in the 4.2 production cycle then.

YuriSizov commented:
There are still formatting issues which fail CI and it needs a rebase. After that, if you feel confident that this is in a state that is ready for public testing, we can merge it for one of the first dev snapshots.

AThousandShips commented on editor/plugins/script_editor_plugin.h:
```diff
@@ -393,7 +394,10 @@ class ScriptEditor : public PanelContainer {
+	bool format_on_save;
+	bool use_space_indentation;
 	bool convert_indent_on_save;
+	void _format_code(ScriptEditorBase *se);
```
```suggestion
	void _format_code(ScriptEditorBase *p_se);
```
Prefix arguments with `p_`

AThousandShips commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -1008,6 +1000,23 @@ void ScriptEditor::_scene_saved_callback(const String &p_path) {
 	}
 }
 
+void ScriptEditor::_format_code(ScriptEditorBase *se) {
```
```suggestion
void ScriptEditor::_format_code(ScriptEditorBase *p_se) {
```

aaronfranke replied:
`p_script_editor` would be even better.

AThousandShips replied:
Ping for this one

NathanLovato commented:
A quick note as we haven't been able to continue work on this for two months: we absolutely needed Adam to work on educational tech and address some HTML5 issues in Godot 4, aside from his work in the production team.

This particular tech (codepen-like for Godot 4) was a requirement for our new course, which is also the foundation of the free education program we're making for public schools. We have partner schools that are going to use this by the end of the year, so it couldn't be delayed.

This is just done now since Friday, so Adam will be able to resume work on the formatter and complete this contribution soon.

adamscott committed:
Rename variable to avoid shadowing another one

adamscott committed:
Apply code style suggestions

adamscott commented:
Thanks @AThousandShips for the code style review! It was needed.

adamscott committed:
Apply misc/scripts/header_guards.sh

adamscott committed:
Fix StringBuilder import missing

adamscott committed:
Fix crash when begining to write a match statement

adamscott committed:
Fix issues with binary operators indent

adamscott committed:
Swap pre_formatted and output

adamscott committed:
Fix array indent issue

adamscott committed:
Fix dictionary indent

adamscott committed:
Fix parameter variable name

adamscott commented:
Make sure to check if it impacts exported code.

SlugFiller commented:
I don't think a "Comment" token should be added, because it makes work for the parser harder, because it now needs to manually skip/handle comments.

It also has the potential to break things. For example:
```gdscript
func MyFunc():
  action1()
# Non-indented comment
  action2()
```
The comment is not indented, but it doesn't cause a DEDENT, and therefore, does not cause an error with relation to the function's indentation.

An alternative is to instead turn comments into metadata that is stored in the next token (or the end-of-file token). The parser just sees a normal token, but can retain a reference to it when building the tree. Then, the formatter can read back the comments when reading the AST. It can then try to output the comment as-is, or make whatever modifications are necessary to avoid breaking the code.

A similar handling can be applied to whitespace. In fact, comments and whitespace should be treated as a single unit.

A decent alternative to putting the comment on an arbitrary token, is to put it in the nearest whitespace token, like NEWLINE, INDENT, DEDENT, EOF. Since a comment can't be in the middle of a line, only at the end of one. Basically, you can store their original text, including both whitespace and comments, in the same way that identifiers store their name.

Also, proper use of `#ifdef` should be used to ensure that this isn't include in the export templates, because games don't need to reformat their scripts, but they do need to parse them.

NathanLovato commented:
This PR needs discussion and a thorough review at this stage. Comments were already discussed with the GDScript contributors when first implemented, I believe the current implementation resulted from that, and from the request from users to have comments placed pretty precisely when formatting.

I'll let Adam, who's now in charge of the code, detail the challenges he's facing to wrap this up. On our end, we're running out of money, so at least for the foreseeable future (until our finances are back in the green), we will only be able to pay Adam for another few days on this.

If the project requires more than that to reach a merging state, we'll need either a sponsorship or help from other developers. Or time - we want to help complete this, but we need to have the budget for it.

dalexeev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -489,14 +583,27 @@ void GDScriptParser::parse_program() {
 				if (annotation->applies_to(AnnotationInfo::SCRIPT)) {
-					// `@icon` needs to be applied in the parser. See GH-72444.
+					// `@icon` and `@tool` need to be applied in the parser. See GH-72444.
 					if (annotation->name == SNAME("@icon")) {
+						head->icon_inline_comment = check_for_comment();
+						match(GDScriptTokenizer::Token::NEWLINE);
+						head->icon_header_comment = top_header_comment.duplicate();
+						top_header_comment.clear();
+						annotation->apply(this, head);
+					} else if (annotation->name == SNAME("@tool")) {
+						head->tool_inline_comment = check_for_comment();
+						match(GDScriptTokenizer::Token::NEWLINE);
+						head->tool_header_comment = top_header_comment.duplicate();
+						top_header_comment.clear();
 						annotation->apply(this, head);
```
Can this be done universally, without separate properties for individual annotations? Perhaps in the future we will need other annotations (#76020), which should be resolved in the parser.

adamscott committed:
Fix some issues

AThousandShips commented:
We are currently in feature freeze until 4.2 is released so no changes can be merged like this, only then can this be considered and merged, so come back and see when 4.2 has been released 🙂

Calinou commented:
Tested locally with https://github.com/godotengine/godot-demo-projects/tree/2d/platformer, it mostly works as expected. All files format without errors in that demo, and the demo keeps working after having all its scripts formatted.

### Questions

- Is there an annotation or comment to avoid formatting the next line, or within a series of lines (like `// clang-format` comments with begin and end markers)?

### Bugs

- Running formatter twice sometimes results in different code (that is still valid, but nonetheless different):

#### No formatter

```gdscript
if mode == DisplayServer.WINDOW_MODE_FULLSCREEN or \
				mode == DisplayServer.WINDOW_MODE_EXCLUSIVE_FULLSCREEN:
```

- The formatter sometimes inserts more than one blank line *within a function*, which isn't allowed as per the GDScript style guide:

```gdscript
	if not floor_detector_left.is_colliding():
		velocity.x = WALK_SPEED
	elif not floor_detector_right.is_colliding():
		velocity.x = -WALK_SPEED

	if is_on_wall():
		velocity.x = -velocity.x


	move_and_slide()  # ^ Two blank lines above ^

	if velocity.x > 0.0:
		sprite.scale.x = 1.0
	elif velocity.x < 0.0:
		sprite.scale.x = -1.0
```

#### Formatted once

```gdscript
		if (
			mode
			== DisplayServer.WINDOW_MODE_FULLSCREEN
			or mode
			== DisplayServer.WINDOW_MODE_EXCLUSIVE_FULLSCREEN
		):
```

#### Formatted twice

```gdscript
		if (
			(
				mode
			== DisplayServer.WINDOW_MODE_FULLSCREEN
			or mode
			== DisplayServer.WINDOW_MODE_EXCLUSIVE_FULLSCREEN
			)
		):
```

- Running formatter on 3D platformer's `player.gd` results in a script error due to incorrect formatting:

```gdscript
facing_mesh = (facing_mesh - Vector3.UP * facing_mesh.dot(Vector3.UP)).normalized()
```

This is because the formatter gets rid of parentheses too aggressively in this situation:

```gdscript
facing_mesh = facing_mesh - Vector3.UP * facing_mesh.dot(Vector3.UP).normalized()
```

### Suggestions for improving formatting

- Both operands within a comparison operation should always be written on the same lime, regardless of length. This looks strange even though it's valid GDScript:

```gdscript
		if (
			mode
			== DisplayServer.WINDOW_MODE_FULLSCREEN
			or mode
			== DisplayServer.WINDOW_MODE_EXCLUSIVE_FULLSCREEN
		):
```

I'd prefer the following:

```gdscript
		if (
			mode == DisplayServer.WINDOW_MODE_FULLSCREEN
			or mode == DisplayServer.WINDOW_MODE_EXCLUSIVE_FULLSCREEN
		):
```

- Spacing between `extends` and variables, and variables and the first function should be 1 line only, not 2. There isn't much of a benefit of using 2-line spacing in this context:

```gdscript
class_name Game
extends Node


@onready var _pause_menu := $InterfaceLayer/PauseMenu as PauseMenu


func _unhandled_input(event: InputEvent) -> void:
```

I'd expect this instead:

```gdscript
class_name Game
extends Node


@onready var _pause_menu := $InterfaceLayer/PauseMenu as PauseMenu


func _unhandled_input(event: InputEvent) -> void:
	pass
```

- The formatter will sometimes reorganize comments placed at the end of a block to put them at the beginning of the next block, which can be confusing:

```gdscript
	... 
	elif col_left.is_empty() and not col_right.is_empty():
		# If only right ray is occluded, turn the camera around to the left.
		difference = Basis(Vector3.UP, deg_to_rad(delta * autoturn_speed)) * difference
	# Do nothing otherwise, left and right are occluded but center is not, so do not autoturn.

	# Apply lookat.
	if difference.is_zero_approx():
		difference = (pos - target).normalized() * 0.0001
```

Becomes:

```
	... 
	elif col_left.is_empty() and not col_right.is_empty():
		# If only right ray is occluded, turn the camera around to the left.
		difference = Basis(Vector3.UP, deg_to_rad(delta * autoturn_speed)) * difference

	# Do nothing otherwise, left and right are occluded but center is not, so do not autoturn.
	# Apply lookat.
	if difference.is_zero_approx():
		difference = (pos - target).normalized() * 0.0001
```

- The 2D platformer demo currently uses single-line class name declarations like this: `class_name Enemy extends CharacterBody2D`, which reminds me of C++/Java. This saves one line in every named class script without much of a downside (as this line is usually not wide enough to warrant splitting into two lines). The formatter always reformats this to two lines, but this would need to be discussed separately as the current style guide also disagrees.
- The formatter doesn't do anything about signals that have no parameters but parentheses like this: `signal coin_collected()`. We should discuss whether the parentheses should be enforced or forbidden by the style guide, and make the formatter act accordingly. (I'd go towards enforcing them personally, like functions.)
- The formatter should add 2 spaces before inline comments: `var something  # Like this`, instead of just 1. This is consistent with what Python formatters do.

### Other suggestions

- I think the default keyboard shortcut should be <kbd>Ctrl + Shift + I</kbd> for consistency with VS Code. I got pretty used to that shortcut for formatting over the years 🙂 
  - This is also because keyboard shortcuts that use <kbd>Alt + Shift</kbd> aren't very common, and they cause issues on Windows where you accidentally switch keyboard layouts while trying to enter them (<kbd>Alt + Shift</kbd> cycles keyboard layouts on Windows).
- There should be a `--format` CLI argument for usage on CI, analogous to `--check-only`. Returns exit code `0` ("success") if no file was reformatted, `1` ("error") if at least one file was reformatted, so that CI can exit with a failure if files don't follow formatting standards set in the project.
  - This argument should be recursive if you specify a folder name. A top-level `addons/` folder (i.e. next to a `project.godot` file) should be ignored unless you explicitly specify it by name, so that formatting errors in add-ons don't impact your project. Specifying by name should still be allowed so that add-ons themselves can use automated formatting.

emgast commented:
Is this feature still being actively developed? I'd love to see it get into 4.3, but I don't know how realistic that is. Either way, I really appreciate the effort!

aaronfranke commented:
@emgast It is not being actively developed, the last time changes were pushed was 4 months ago.

AThousandShips commented:
It needs a rebase, might need some new fixes

NathanLovato commented:
I can confirm it is currently not in active development. Adam, the last developer we sponsored to work on this, was rehired by the Godot foundation, and he's now working on higher priority features and fixes for the community. The groundwork is here, with many automated tests. We ran out of budget for this, now we'll need a new contributor or sponsor willing to pick it up and complete this.

AThousandShips commented:
I will take a look over this when I have the time and might be able to salvage it, but will see if I can get my head around all the details

NathanLovato commented:
If anyone gives it a stab, don't hesitate to ask Adam, he can surely give you a quick overview of how the thing works and the areas that caused issues.

Mainly, from what I understood, the current implementation faces the same challenges as any formatter that tries to output nice-looking code (like Python's black, the reference for this one): line splitting is complicated, and the algorithm might need a rework.

Then, there are lots of small edge cases to address, many of which were listed by Calinou above.

Scony commented:
Not sure what was the reasoning behind implementing this feature as a single PR, but IMO it's virtually impossible to finish it this way - this is basically because GDScript internals are likely changing much faster than this PR can be aligned by a single person.

What I would suggest instead is implementing formatter in small steps (PRs up to 200-300LOC of code + tests) - one part at a time.

The first PR would basically:
- introduce a "dummy" formatter that can format empty file or a file containing just some simple statements like `pass`.
- introduce a means to run the formatter e.g. by using the keyboard shortcut in the code editor yet with the limitation such that formatter should only work if some Godot setting (disabled by default) is enabled. This way, just after the first PR people would be able to enable such "experimental" formatter at their own risk and start testing.
- introduce test cases covering whatever is implemented in the formatter at this point.

Every consecutive PR would add one small functionality along with tests covering it.

The benefit of such "baby steps" approach would be that:

1. PRs are easy to review since they are small
2. More people would be able to get involved into development
3. More people would be able to get involved into testing
4. **Most important!** Every change in the GDScript internals that is breaking formatter would make formatter unit/integration tests failing thus forcing people working on GDScript internals to fix formatter in the same PR they're making changes. Example: _Imagine the formatter is able to format `pass` statements and has unit tests for that. Now imagine some contributor is changing `pass` statement to `noop` statement for whatever reason. PR with such change would fail formatter tests immediately thus forcing the contributor to fix formatter before committing such PR_

So, in other words, working in the `master` directly would enable organic development of formatter and make it a team effort - especially on the development side.

NathanLovato commented:
@Scony initially, the development was staged like that with a simpler formatter that would focus on the foundations. It couldn't get merged because to preserve user comments, they need to be added to the parser somehow, and from that point on the formatter becomes a project that adds a change to the GDScript parser and needs to be thoroughly tested before merging. The brief was that it generally shouldn't break any code we know before being exposed to users. 

It's great if the team can find a way to split the work now. For instance, by keeping the formatter unexposed to users.

Anyway, this code belongs to the project, we did what we could, and I'll let the GDScript team decide on where to take it from there. You're more than welcome to split it, change it, discard it, or anything else you need to do.

vnen commented:
I agree with @Scony's idea of doing it in small steps, but I don't know if I agree on what the steps should. IMO the first step would be a "noop" formatter that is able to read to the code and leave it exactly as it was found. Then start adding rules one by one, always backed by automated tests.

Doing a noop is much harder than formatter that only does one rule and discards the rest, but we can get that into production faster even without all the rules.

Not sure if @adamscott is still interested in this. He could work on this as part of his tasks within the Godot Foundation if his schedule allows for it.

chexmo commented:
Sorry in advance for being new in these repos and be trying to quote stuff.... You can dismiss this message if it's garbage.

I was thinking... If both the parser and the formatter need to follow "code rules"... isn't it a chance to make "code rules" an actual system that them both can connect to?

I example: A code rule may be "the sign of a method should look like this", and both the parser and the formatter (even though they are pretty different functionalities in the user's eyes) can read the rule and then do what they need to do depending the tool selected and their configuration

The benefit from this is that you can add and delete code rules when the GDScript specification changes without touching the tools that use them.

khellste commented:
@NathanLovato is it out of the question to simply expose `GDScriptParser` as an API accessible to plugin scripts. Doing so could let plugin/add-on authors make their own formatters without having to reimplement the GDScript parser.

NathanLovato commented:
@khellste this is not something I couldn't answer. I wouldn't do this for the purpose of making formatters, however:

- The GDScript parser, as it stands, is not designed to support a fully-fledged formatter because it only captures tokens that are relevant to the code's execution and discards the rest.
- The bulk of the code and challenge in a formatter is generally making the formatter.
- There are third-party parsers that could be used instead.

Existing GDScript parsers:

1. The Python-based parser by Scony, which also comes with a formatter: https://github.com/Scony/godot-gdscript-toolkit
2. The tree sitter parser, which is used by code editors like neovim and zed: https://github.com/PrestonKnopp/tree-sitter-gdscript

For making a new third-party parser, Adam Scott suggested looking into the latter and a library called topiary (https://github.com/tweag/topiary) which is designed to make creating a simple formatter easy based on an existing tree sitter parser. It would likely be the fastest route right now to get something production-ready.

NathanLovato commented:
@khellste I don't think the approach used in this PR will likely get merged. Someone could get a proposal up and see if it catches on but... I don't think it's gonna be a priority for anyone to look at something like this soon. It'd be way faster to fork and make the changes you need for yourself.

Scony was looking to take over the project and salvage some of the code starting in https://github.com/godotengine/godot/pull/97383 (making the official formatter with small incremental PRs with the hope they'll get reviewed and merged more easily), and from what I understood, make the parsing bits that are unique to the needs of a formatter part of the formatter.

To me, this is our best bet to get a GDScript formatter, especially as he's made the community one - he has experience specifically with making a GDScript formatter. But this depends on maintainers being available to review, confirm the direction, and merge his work. It'll probably take time.

khellste commented:
@NathanLovato good to know, I wasn't aware of #97383. Scony seems like a natural choice for it given his background. I'll redirect my attention to the new PR/project if I have any questions or ideas.

Thanks for taking the time to reply to me in this thread!

Scony commented:
> @NathanLovato good to know, I wasn't aware of #97383. Scony seems like a natural choice for contributing a formatter given his background. I'll redirect my attention to the new PR/project if I have any questions or ideas.
> 
> Thanks for taking the time to reply to me in this thread!

The https://github.com/godotengine/godot/pull/97383 is technically ready to merge but it looks like we must wait a bit to let that sink in. Anyway - once this is merged, a series of follow-up PRs will be done. I'll try to make them referring to https://github.com/godotengine/godot/pull/97383 so that you can use https://github.com/godotengine/godot/pull/97383 for tracking the progress.
